### PR TITLE
Small fixes to github-linguist command line tool when used in single file mode.

### LINKED ISSUE #5381: github-linguist ignores .gitattributes when used in single-file mode
## Describe the bug

In my `.gitattributes` file I have;
```
checks/tests/license/test-missing-spdx-generated.v linguist-generated
```
git reports this attribute is set correctly;
```
$ git check-attr --all checks/tests/license/test-missing-spdx-generated.v
checks/tests/license/test-missing-spdx-generated.v: linguist-generated: set
```

However, `github-linguist` doesn't report this file as being generated when I use;
```
$ github-linguist checks/tests/license/test-missing-spdx-generated.v
test-missing-spdx-generated.v: 5 lines (3 sloc)
  type:      Text
  mime type: text/plain
  language:  Verilog
```

This appears to be caused by the fact that when `github-linguist` is operating in single file mode it ignores that the file is inside a git repository.

### Expected behaviour

Single file mode works like the multifile mode.

### PR DESCRIPTION
**Pull request by mithro**
## Description

The `github-linguist` tool is useful for examining how linguist thinks about your repository but has a number of small issues when given a single file name. This pull request fixes these problems.

Fixes #5381, #5380 and #5379

### CODE REVIEW

**Comment by mithro** [comment]
I'm not a ruby programmer, so I don't know if I'm doing things here in the correct "idiomatic" ruby way.

**Comment by mithro** [comment]
FYI - @acomodi

**Comment by lildude** [comment]
Thanks for this. I've been meaning to do this myself after a recent [discussion](https://github.com/github/linguist/discussions/5308) which brought up the limitations. 

> I'm not a ruby programmer, so I don't know if I'm doing things here in the correct "idiomatic" ruby way.

I'm happy to help you here and contribute changes directly to your PR. If you're happy with this, please grant me write access to your PR and I'll commit my changes directly.

I'll also need to check the behaviour against the main GitHub.com code before merging as I know several places use Linguist in different forms so I need to be sure nothing breaks.

**Review by Alhadis — COMMENTED** [review]
Something that's always bugged me: can't Linguist analyse more than one file? Output could be separated by a blank line, or possibly the filepath (similar to `exiftool`'s behaviour).

The primary benefit, other than adherence to Unix philosophy and user expectations, would be a drastic performance boost over a shell loop to achieve the equivalent:

```console
$ for i in "$@"; do github-linguist "$i"; done
```

Since every time the Linguist executable is run, it has to keep reinitialising an entire Ruby runtime, packed with a simples and language database...

**Comment by mithro** [comment]
"Allow edits and access to secrets by maintainers" is already set, so please go ahead and help fix things up.

**Comment by mithro** [comment]
What can I do to get this moving forward?

**Comment by lildude** [comment]
Nothing right now, I'm incredibly busy right now so haven't found the time to take a proper look. I will as soon as I've got the time.

**Comment by lildude** [comment]
@mithro I've just pushed my refactor. I've switched to using more descriptive variable names, removed an unnecessary variable assignment and have switched out the single-use method for a lambda function as I think you only went for the method so you can exit early.

**Comment by mithro** [comment]
Looks good to me!

**Review by lildude — APPROVED** [review]
Tested against our internal use and no issues were found. Thanks for this improvement.

**Comment by lildude** [comment]
Oooof! I've just discovered this breaks running Linguist using `bundle exec` where you need to provide a full path to the file:

```console
$ bundle exec bin/github-linguist ~/tmp/trash/overmind/dialer.go
bundler: failed to load command: bin/github-linguist (bin/github-linguist)
Traceback (most recent call last):
	16: from /Users/lildude/.asdf/installs/ruby/2.7.2/bin/bundle:23:in `<main>'
	15: from /Users/lildude/.asdf/installs/ruby/2.7.2/bin/bundle:23:in `load'
	14: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/exe/bundle:37:in `<top (required)>'
	13: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/friendly_errors.rb:130:in `with_friendly_errors'
	12: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/exe/bundle:49:in `block in <top (required)>'
	11: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/cli.rb:24:in `start'
	10: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
	 9: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/cli.rb:30:in `dispatch'
	 8: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
	 7: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	 6: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	 5: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/cli.rb:494:in `exec'
	 4: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/cli/exec.rb:28:in `run'
	 3: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/cli/exec.rb:63:in `kernel_load'
	 2: from /Users/lildude/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/bundler-2.2.17/lib/bundler/cli/exec.rb:63:in `load'
	 1: from bin/github-linguist:57:in `<top (required)>'
bin/github-linguist:57:in `discover': failed to resolve path '/Users/lildude/github/linguist/Users/lildude/tmp/trash/overmind/dialer.go': No such file or directory (Rugged::OSError)
$
```

Note the double `/Users/lildude/github/linguist/` in the path in the last line of the error.

Determining the full path is assuming the path provided is relative to the current location, which isn't the case when using `bundle exec`. The same would apply with normal `github-linguist` when a new release is made.

**Pull request merged**
